src/rest-server/src/middlewares/v2/protocol.js

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+// module dependencies
+const yaml = require('js-yaml');
+const mustache = require('mustache');
+const createError = require('../../util/error');
+const protocolSchema = require('../../config/v2/protocol');
+
+
+const prerequisiteTypes = [
+  'script',
+  'output',
+  'data',
+  'dockerimage',
+];
+
+const prerequisiteFields = [
+  'script',
+  'output',
+  'data',
+  'dockerImage',
+];
+
+const protocolWrap = async (req, res, next) => {
+  const protocolYAML = req.body;
+  req.rawBody = protocolYAML;
+  req.body = {
+    protocol: protocolYAML,
```

Comment:
In case we may need to add more props in the body.

YAML request body can be parsed the same as a text request by `express`.
raw YAML text is stored in `rawBody` in current implementation.

Reply:
Understand, and `req.body` should always just hold the body of the request, whatever it is text, json or yaml.

If you have any other props related to the request body, here are two ways:

1. Store it in another req props, especially `req.local`.
2. Explicitly make another class / type like "protocol", and treat the middleware as "prototol-parser".

Reply:
updated, stored the rendered js object in another prop `req.protocol`

Reply:
Shadowed [req.protocol](https://expressjs.com/en/4x/api.html#req.protocol)...